src/vs/editor/contrib/colorPicker/browser/colorDetector.ts

```diff
@@ -58,28 +59,33 @@ export class ColorDetector extends Disposable implements IEditorContribution {
 		super();
 		this._debounceInformation = languageFeatureDebounceService.for(_languageFeaturesService.colorProvider, 'Document Colors', { min: ColorDetector.RECOMPUTE_TIME });
 		this._register(_editor.onDidChangeModel(() => {
-			this._isEnabled = this.isEnabled();
-			this.onModelChanged();
+			this._isColorDecoratorsEnabled = this.isEnabled();
+			this.updateColors();
 		}));
-		this._register(_editor.onDidChangeModelLanguage(() => this.onModelChanged()));
-		this._register(_languageFeaturesService.colorProvider.onDidChange(() => this.onModelChanged()));
+		this._register(_editor.onDidChangeModelLanguage(() => this.updateColors()));
+		this._register(_languageFeaturesService.colorProvider.onDidChange(() => this.updateColors()));
 		this._register(_editor.onDidChangeConfiguration((e) => {
-			const prevIsEnabled = this._isEnabled;
-			this._isEnabled = this.isEnabled();
-			const updated = prevIsEnabled !== this._isEnabled || e.hasChanged(EditorOption.colorDecoratorsLimit);
-			if (updated) {
-				if (this._isEnabled) {
-					this.onModelChanged();
-				} else {
+			const prevIsEnabled = this._isColorDecoratorsEnabled;
+			this._isColorDecoratorsEnabled = this.isEnabled();
+			const prevIsDefaultColorsEnabled = this._isDefaultColorDecoratorsEnabled;
+			this._isDefaultColorDecoratorsEnabled = this._editor.getOption(EditorOption.defaultColorDecorators);
+			const updatedColorDecoratorsSetting = prevIsEnabled !== this._isColorDecoratorsEnabled || e.hasChanged(EditorOption.colorDecoratorsLimit);
+			const updatedDefaultColorDecoratorsSetting = prevIsDefaultColorsEnabled !== this._isDefaultColorDecoratorsEnabled || e.hasChanged(EditorOption.defaultColorDecorators);
+			if (updatedColorDecoratorsSetting || updatedDefaultColorDecoratorsSetting) {
+				if (!this._isColorDecoratorsEnabled) {
 					this.removeAllDecorations();
 				}
+				// Update the color data in all cases so the color picker can use the default document color provider
+				// even when the decorators are disabled.
```

Comment:
Right now, even if `_isColorDecoratorsEnabled` is false, all providers and the default provider are computing colors. No color boxes are rendered, but editor decorations are created. The editor decorations are stored in `_colorDatas` (line 194) and are used for `getColorData`. `getColorData` is used by the hover.

I would suggest to not compute anything `_isColorDecoratorsEnabled` is false
- color decorator computation is quite exensive. Computation can involve language extensions on the (potentially remote) extension host, that can also be remote.
- there should no hover if there are no color boxes
- if the information is used for the `show color picker` command: it's better to compute the color ranges on demand in that case. I don't think that command is called so ofter to justfy computing color ranges constantly.
- also there's a delay until the color detector has computed ranges and created decorations. Given that, I would even say the `show color picker` command should always compute color decorations new.